src/pages/Cart/__test__/Cart.test.js

```diff
@@ -0,0 +1,109 @@
+
+describe("Shopping Cart Empty State", () => {
+  it("should display an empty state when there are no items in the cart", async () => {
+    renderWithRouterAndRedux(<App />, {
+      route: `/cart`,
+    });
+
+    expect(screen.getByText(/Shopping Cart is Empty/i)).toBeInTheDocument();
+  });
+});
+
+describe("Shopping Cart with initial state", () => {
+  beforeEach(() => {
+    renderWithRouterAndRedux(<App />, {
+      route: `/cart`,
+      initialState: initialState,
+    });
+  });
+
+  it("should validate the list of the products is shown when there are items in the cart", () => {
+    const products = initialState.cart.products;
+    const { format } = formatter;
+
+    for (const product of products) {
+      const price = format(product.data.price);
+      const subtotal = format(product.data.price * product.quantity);
+
+      expect(screen.getByText(product.data.name)).toBeInTheDocument();
+      expect(screen.getByTestId(`price-${product.id}`)).toHaveTextContent(
+        price
+      );
+      expect(screen.getByTestId(`subtotal-${product.id}`)).toHaveTextContent(
+        subtotal
+      );
+      expect(
+        screen.getByTestId(`remove-icon-${product.id}`)
+      ).toBeInTheDocument();
+      expect(screen.getByTestId(`quantity-${product.id}`)).toBeInTheDocument();
+    }
+  });
+
+  it("should validate the cart total label display the sum of the subtotals of all items in the cart", () => {
+    const { format } = formatter;
+    const cartTotal = format(initialState.cart.total);
+
+    expect(screen.getByText(`Cart Total: ${cartTotal}`)).toBeInTheDocument();
+  });
+
+  it("should update the total items in the cart when using the quantity selector for a particular product and validate the out of stock", () => {
+    const product = initialState.cart.products[0];
+    const stock = product.data.stock;
+    const currQuantity = product.quantity;
+    const countItems = initialState.cart.count;
+
+    const increment = screen.getByTestId(`increment-${product.id}`);
+    const totalCount = screen.getByTestId("total-count");
+
+    expect(increment).not.toBeDisabled();
+    expect(totalCount).toHaveTextContent(`(${countItems})`);
+
+    //Act increment quantity
+    for (let index = currQuantity; index < stock; index++) {
+      fireEvent.click(increment);
+    }
+
+    const newCountItems = countItems + (stock - currQuantity);
+
+    expect(increment).toBeDisabled(); //Increment button disabled
+    expect(totalCount).toHaveTextContent(`(${newCountItems})`); //New count total items
+  });
+
+  it("should validate that a product can be removed after clicking on the remove icon", () => {
+    const product = initialState.cart.products[0];
+    const removeIcon = screen.getByTestId(`remove-icon-${product.id}`);
+
+    expect(screen.getByText(product.data.name)).toBeInTheDocument();
+
+    fireEvent.click(removeIcon);
+
+    expect(() => screen.getByText(product.data.name)).toThrow(
+      "Unable to find an element"
+    );
```

Comment:
There's an option here that you could use. Instead of "getting" by text, you should be "querying" by text instead (`queryByText`). The difference between get and query is that query won't `throw` if the element is not found, it will return null. When using "get" you let the testing library that the element is in there, and you want it. The query one is when you are no sure if it is in there or not.

Then use `.not.toBeInTheDocument();` instead of checking for it to throw. Like so:

```javascript
expect(screen.queryByText(product.data.name)).not.toBeInTheDocument();
```

Reply:
Thanks, I changed all the toThrow expects for not.toBeInTheDocument, it's more readable.